lib/racecar/producer.rb

```diff
@@ -0,0 +1,96 @@
+# frozen_string_literal: true
+
+require "racecar/message_delivery_error"
+
+
+module Racecar
+  class Producer
+
+    def initialize(config: nil, logger: nil, instrumenter: NullInstrumenter)
+      @config = config
+      @logger = logger
+      @delivery_handles = []
+      @instrumenter = instrumenter
+      @batching = false
+    end
+
+
+    # fire and forget - you won't get any guarantees or feedback from 
+    # Racecar on the status of the message and it won't halt execution
+    # of the rest of your code.
+    def produce_async(value:, topic:, **options)
+      handle = internal_producer.produce(payload: value, topic: topic, **options)
+      @delivery_handles << handle if @batching
+
+      nil
+    end
+
+    # synchronous message production - will wait until the delivery handle succeeds, fails or times out.
+    def produce_sync(value:, topic:, **options)
+      begin
+        handle = internal_producer.produce(payload: value, topic: topic, **options).wait(max_wait_timeout: 60, wait_timeout: 0.1)
+      rescue Rdkafka::AbstractHandle::WaitTimeoutError => e
+        partition = MessageDeliveryError.partition_from_delivery_handle(handle)
+        @logger.debug "Still trying to deliver message to (partition #{partition})... (will try up to Racecar.config.message_timeout)"
+        retry
+      rescue Rdkafka::RdkafkaError => e
+        raise MessageDeliveryError.new(e, handle)
+      end
+      nil
+    end
+
+    # Blocks until all messages that have been asynchronously produced in the block have been delivered.
+    # Usage:
+    # messages = [
+    #             {value: "message1", topic: "topic1"},
+    #             {value: "message2", topic: "topic1"},
+    #             {value: "message3", topic: "topic2"}
+    #             ]
+    # Racecar.wait_for_delivery {
+    #   messages.each do |msg|
+    #     Racecar.produce_async(value: msg[:value], topic: msg[:topic])
+    #   end
+    # }
+    def wait_for_delivery
+      @batching = true
+      @delivery_handles.clear
+      yield
+      instrumentation_payload = { delivered_message_count: @delivery_handles.size }
+      @instrumenter.instrument('wait_for_delivery', instrumentation_payload) do
+        @delivery_handles.each do |handle|
+          begin
+            handle.wait
+          rescue Rdkafka::AbstractHandle::WaitTimeoutError => e
+            partition = MessageDeliveryError.partition_from_delivery_handle(handle)
+            @logger.debug "Still trying to deliver message to (partition #{partition})... (will try up to Racecar.config.message_timeout)"
+            retry
+          rescue Rdkafka::RdkafkaError => e
+            raise MessageDeliveryError.new(e, handle)
+          end
```

Comment:
We can avoid this duplication using a helper method, e.g.

```ruby
private
def deliver_with_error_handling(handle)
  handle.wait
rescue Rdkafka::AbstractHandle::WaitTimeoutError => e
  partition = MessageDeliveryError.partition_from_delivery_handle(handle)
  @logger.debug "Still trying to deliver message to (partition #{partition})... (will try up to Racecar.config.message_timeout)"
  retry
rescue Rdkafka::RdkafkaError => e
  raise MessageDeliveryError.new(e, handle)
end
```

Reply:
Thanks - I will add that and use it in `produce_sync` as well since `rdkafka-ruby`'s `wait` is using the same values as [us per default](https://github.com/appsignal/rdkafka-ruby/blob/c02f217a189b71e3e253fe4ad59286b5fc9d4034/lib/rdkafka/abstract_handle.rb#L45-L63):

```
def wait(max_wait_timeout: 60, wait_timeout: 0.1)
      timeout = if max_wait_timeout
                  CURRENT_TIME.call + max_wait_timeout
                else
                  nil
                end
      loop do
        if pending?
          if timeout && timeout <= CURRENT_TIME.call
            raise WaitTimeoutError.new("Waiting for #{operation_name} timed out after #{max_wait_timeout} seconds")
          end
          sleep wait_timeout
        elsif self[:response] != 0
          raise_error
        else
          return create_result
        end
      end
    end
```

Reply:
We can create a seperate PR for allowing users to set the values to whatever they want.